### PR TITLE
VZ-5337.  Explicitly reject any post-install updates that have not been vetted.

### PR DESCRIPTION
**Pull request by michael-cico**
# Description

Update the component-level validators to reject any edits that have not been validated.  These will be relaxed over time as we verify what updates are possible.

Fixes VZ-5337.

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/authproxy/authproxy_component.go [hunk 1]
```diff
@@ -68,6 +69,16 @@ func (c authProxyComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.Ver
 	if c.IsEnabled(old) && !c.IsEnabled(new) {
 		return fmt.Errorf("can not disable previously enabled %s", ComponentJSONName)
 	}
+	if !reflect.DeepEqual(c.getKubernetesSettings(old), c.getKubernetesSettings(new)) {
```
Comment (mgianatagh) [inline platform-operator/controllers/verrazzano/component/authproxy/authproxy_component.go:72]: One of the tests I tried when update functionality was added was to increase the replica count of authproxy.  With this change, would an update to the CR get rejected if changing replica count?

File: platform-operator/controllers/verrazzano/component/nginx/nginx_component.go [hunk 2]
```diff
@@ -76,5 +79,26 @@ func (c nginxComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.Verrazz
 	if c.IsEnabled(old) && !c.IsEnabled(new) {
 		return fmt.Errorf("can not disable previously enabled %s", ComponentJSONName)
 	}
+	oldType, _ := vzconfig.GetServiceType(old)
+	newType, _ := vzconfig.GetServiceType(new)
+	if !reflect.DeepEqual(c.getInstallArgs(old), c.getInstallArgs(new)) ||
+		!reflect.DeepEqual(c.getPorts(old), c.getPorts(new)) ||
+		oldType != newType {
+		return fmt.Errorf("Update not allowed for %s", ComponentJSONName)
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/nginx/nginx_component.go:87]: might be better Deep equal each field and tell user exactly what field(s) cannot be changed.

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go [hunk 3]
```diff
@@ -166,12 +180,88 @@ func (c verrazzanoComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.Ve
...
+	}
+	if !reflect.DeepEqual(getESInstallArgs(old), getESInstallArgs(new)) {
+		return fmt.Errorf("Opensearch updates not allowed for %s", ComponentJSONName)
+	}
+	return nil
+}
+
+func compareFluentd(old *vzapi.Verrazzano, new *vzapi.Verrazzano) error {
+	if vzconfig.IsFluentdEnabled(old) && !vzconfig.IsFluentdEnabled(new) {
+		return fmt.Errorf("can not disable Fluentd when previously enabled %s", ComponentJSONName)
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go:204]: First work in all messages should be capitalized.  Please do this throughout.

File: platform-operator/controllers/verrazzano/component/nginx/nginx_component.go [hunk 4]
```diff
@@ -76,5 +79,26 @@ func (c nginxComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.Verrazz
 	if c.IsEnabled(old) && !c.IsEnabled(new) {
 		return fmt.Errorf("can not disable previously enabled %s", ComponentJSONName)
 	}
+	oldType, _ := vzconfig.GetServiceType(old)
+	newType, _ := vzconfig.GetServiceType(new)
+	if !reflect.DeepEqual(c.getInstallArgs(old), c.getInstallArgs(new)) ||
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/nginx/nginx_component.go:84]: I am confused about what we are supposed to be allowing/not allowing.  Can you add a comment for each component saying what can be changed?

This logic doesn't allow changes to the component at all does it?

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager_component.go [hunk 5]
```diff
@@ -70,5 +71,16 @@ func (c certManagerComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.V
 	if c.IsEnabled(old) && !c.IsEnabled(new) {
 		return fmt.Errorf("can not disable previously enabled %s", ComponentJSONName)
 	}
+	if !reflect.DeepEqual(c.getCertificateSettings(old), c.getCertificateSettings(new)) {
+		return fmt.Errorf("Updates not allowed for certificate settings not allowed for %s", c.GetJSONName())
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/certmanager/certmanager_component.go:75]: Wording on log message is confusing

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go [hunk 6]
```diff
@@ -66,8 +67,12 @@ func (e externalDNSComponent) IsEnabled(effectiveCR *vzapi.Verrazzano) bool {
 
 // ValidateUpdate checks if the specified new Verrazzano CR is valid for this component to be updated
 func (e externalDNSComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.Verrazzano) error {
-	if e.IsEnabled(old) && !e.IsEnabled(new) {
-		return fmt.Errorf("can not remove OCI from dns")
+	// This is essentially a check for all DNS changes, for now
+	if !reflect.DeepEqual(old.Spec.Components.DNS, new.Spec.Components.DNS) {
+		return fmt.Errorf("Updates not allowed for dns field")
 	}
+	/*if e.IsEnabled(old) && !e.IsEnabled(new) {
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:74]: Commented code left behind

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go [hunk 7]
```diff
@@ -66,8 +67,12 @@ func (e externalDNSComponent) IsEnabled(effectiveCR *vzapi.Verrazzano) bool {
 
 // ValidateUpdate checks if the specified new Verrazzano CR is valid for this component to be updated
 func (e externalDNSComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.Verrazzano) error {
-	if e.IsEnabled(old) && !e.IsEnabled(new) {
-		return fmt.Errorf("can not remove OCI from dns")
+	// This is essentially a check for all DNS changes, for now
+	if !reflect.DeepEqual(old.Spec.Components.DNS, new.Spec.Components.DNS) {
+		return fmt.Errorf("Updates not allowed for dns field")
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:72]: `dns` should be capitalized

File: platform-operator/controllers/verrazzano/component/nginx/nginx_component.go [hunk 8]
```diff
@@ -76,5 +79,26 @@ func (c nginxComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.Verrazz
...
+		!reflect.DeepEqual(c.getPorts(old), c.getPorts(new)) ||
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/nginx/nginx_component.go:85]: Not clear if updated failed due to ports changing or service type changing

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go [hunk 9]
```diff
@@ -166,12 +180,88 @@ func (c verrazzanoComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.Ve
...
 		}
 	}
+	return nil
+}
+
+func compareOpensearch(old *vzapi.Verrazzano, new *vzapi.Verrazzano) error {
+	if vzconfig.IsElasticsearchEnabled(old) && !vzconfig.IsElasticsearchEnabled(new) {
+		return fmt.Errorf("can not disable Opensearch when previously enabled %s", ComponentJSONName)
+	}
 	if getOpenSearchDataNodeStorageOverride(old) != getOpenSearchDataNodeStorageOverride(new) {
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go:190]: Should be relaxed with #2757

**Review by anders-swanson — COMMENTED** [review]
Some checks seem unnecessarily restrictive, but I think that was your intention?

File: platform-operator/controllers/verrazzano/component/authproxy/authproxy_component.go (hunk 1 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/authproxy/authproxy_component.go:72]: I will loosen this to allow replica changes.  After talking with Mark and Kevin we can be a little bit looser with this.

Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/authproxy/authproxy_component.go:72]: Fixed next push.

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager_component.go (hunk 5 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager_component.go:75]: Fixed next push.

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go (hunk 7 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:72]: I think we were using names to match the JSON name of the components/fields for messages thrown from the validator to the user.

@pfmackin @guoyongzhang what do you guys think?

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go (hunk 6 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:74]: I'm going to restore this and put it first for now, the DeepCompare will go away soon.

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go (hunk 9 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go:190]: Removed these checks also in my next push.

File: platform-operator/controllers/verrazzano/component/appoper/app_operator_component.go [hunk 10]
```diff
@@ -118,8 +118,9 @@ func (c applicationOperatorComponent) IsEnabled(effectiveCR *vzapi.Verrazzano) b
 
 // ValidateUpdate checks if the specified new Verrazzano CR is valid for this component to be updated
 func (c applicationOperatorComponent) ValidateUpdate(old *vzapi.Verrazzano, new *vzapi.Verrazzano) error {
+	// Do not allow any changes except to enable the component post-install
 	if c.IsEnabled(old) && !c.IsEnabled(new) {
-		return fmt.Errorf("can not disable previously enabled %s", ComponentJSONName)
+		return fmt.Errorf("Can not disable %s", ComponentJSONName)
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/appoper/app_operator_component.go:123]: I don't think this error is very clear.  Why can't the component be disable, was there some sort of problem?  I think it is better to say "Disabling component xyz is not allowed".  Same for all the other messages.  Not a blocker.

**Review by pfmackin — APPROVED** [review]

**Pull request merged**